

Group nr: 22

Name 1 and CID: Rasmus Bäckström (RASBACKS)

Name 2 and CID: Gabriella Bengtsson (GABBENG)

In [195]:
import numpy as np
import copy
import random
import pandas as pd
from collections import Counter
from sklearn.model_selection import train_test_split
from sklearn.utils import resample
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from mining_world import Environment
from IPython.display import Image


# Introduction

This assignment dels with data structures, K-NN, tree-based classifiers and evaluation of the methods deployed on a fictional scenario.  

## Mining world 

<img src="imgs/poster.png" width="800"/>

## Scenario


Humanity has now reached a point where we need to extract and refine more Copium, a precious resource with great value. The only problem is that Copium can only be found on certain uninhabitable planets. This of course means that automated robots are sent instead.      

Copium is naturally very unstable and is only exists very temporary before it decays. There are very specific geological activities and circumstances needed for copium to form. The life cycle of Copium follows. First, a hot stream of liquid magma flows to the surface, creating a hotspot that that looks like a small crater. At the surface, if the conditions are correct, copium can form during the cool-down period. But as stated previously, Copium is unstable in its natural environment and decays to other materials shortly after. 

The formation of these deposits craters are very random, but the heat from them can easilly be detected with a satellite. But there is no way of knowing if the newly formed depoist contains copium from just a satellite, therefor there is a robot rover on the ground with sensors that can collect further measurements. The rover has many sensors that can measure the properties of the ground below it, but of course, Copium can not directly be detected with these types of sensors. This is where the machine learning approach will be used, to take all those measurements and try to classify if the deposit contains Copium or not.  


<img src="imgs/overView.png" width="500"/>


# Navigation - Tree search

This section will show how the naivigation is done. This is not a part of the assignment to understand, but will be used. 

##  Breadth first

The method used is a breadth first search algorithm, it is one of the simplest tree search algorithms and basically tries every option for a fixed number of steps and chooses the best one. The robot only have 4 actions, it can move at every time step in a chosen direction. 

## Actions 

<img src="imgs/actions.png" width="300"/>


In [196]:
class Node():
    def __init__(self, actor):
        self.actor = actor
        self.total_score = 0
    
    def update(self, action, inherited_score):
        score = self.actor.step(action)
        self.total_score = 1.05*inherited_score + score
        return self.total_score
    
    def get_score(self):
        return self.total_score

In [197]:
def breadth_first_search(actor, max_depth, action_space):
    node = Node(copy.deepcopy(actor)) 
    queue_keys = ['0'] # queue to keep track of nodes that has not yet been expanded.  
    visited = {queue_keys[0]: node} # saves visited nodes in order to not recalulate the entire path for each step. 
    
    max_score = -np.inf
    best_action = None

    while True:
        key = queue_keys.pop(0)
        if len(key) > max_depth: # stop at a set depth 
            break    
        node = visited[key]
        
        for action in action_space: # expand all children nodes
            child_node = copy.deepcopy(node)  # copy current node
            score = child_node.update(action=action, inherited_score=node.get_score()) # update node with action
            child_key = key + action # create child node key
            
            if score > max_score: # save best path 
                max_score = score
                best_action = child_key[1]
                
            visited[child_key] = child_node  # add child node to visited nodes.
            queue_keys.append(child_key)  # add child node queue of non expanded nodes. 
            
    return best_action ##


# Collect data

The first step is to collect data that will be used for both training and validation. The available types features can be seen with env.get_sensor_properties() and the actual measurements can be retrieved with env.get_sensor_readings(). It will return a dictionary with the same keys as in env.get_sensor_properties() containg a value for each feature. If the robot is not currently over a deposit, then it will return None. The true label can be extracted with env.get_ground_truth(), which will return a 1 if there is copium in the deposit and 0 if not. 
  

In [198]:
# Have True for first time running and then set to false. 
collect_data = False



In [199]:
env = Environment(map_type=1, fps=500, resolution=(1000, 1000))
env.exit()
sensor_properties = env.get_sensor_properties()
print('Sensor properties', sensor_properties)


Sensor properties ['ground_density', 'moist', 'reflectivity', 'silicon_rate', 'oxygen_rate', 'iron_rate', 'aluminium_rate', 'magnesium_rate', 'undetectable']


In [200]:
if collect_data == True:
    env = Environment(map_type=1, fps=500, resolution=(1000, 1000))
    sensor_properties = env.get_sensor_properties()

    # We can initilize the dictionary the following way.
    data = dict()
    data['copium'] = [] 
    for key in sensor_properties:
        data[key] = []

    for i in range(5000):
        action = breadth_first_search(actor=env.get_actor(), max_depth=3, action_space=env.get_action_space())
        env.step(action)
        # if we are over a deposit. 
        if env.get_sensor_readings() is not None:
            sensor_readings = env.get_sensor_readings()
            copium = env.get_ground_truth()

            for key in sensor_readings:
                data[key].append(sensor_readings[key])
            data['copium'].append(copium)

        env.render()

    env.exit()


# Exerscie 1: Data structure

## a) Pandas data frame 
In this assignment we will work pandas data frame for storing the collected data. First create a pandas data frame from the dictionary. The documentation for it can be found at https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html, only the data feild needs to be filled in with the created dictionary. Call this data frame df. 


In [201]:
# TODO: create pandas dataframe

if collect_data == True:
    df = pd.DataFrame(data= data)


## b) Save data
Collecting data for each time the notebook is opened is time consuming and the tests are not as reproducable if the data is different each time. Therefore we save the dateframe and we can load the data instead. After this you can set collect_data=False above. Use the function pandas function df.to_csv('name_of_file.csv', index=False)

The task is to save the data in a file named mydata.csv

In [202]:

if collect_data == True:
    # TODO: save data to a csv file named mydata.  
    df.to_csv("mydata.csv", index = False)

## c) Load data
The task is to load the data from mydata.csv into a dataframe named df. Use the function pd.read_csv('name_of_file.csv')

In [203]:

# TODO: read data from mydata.csv into a dataframe called df.  
df = pd.read_csv("mydata.csv")

In [204]:
# From the data frame you can access all data for a key with for example:
#print("All data for a feature \n", df["copium"])
#print(df["ground_density"])

# You can access a single sample with:

#print("Single sample from index \n", len(df.iloc[1]))

# You can access all freatures but one with:
#all_features_without_copium = df.drop(columns='copium')


## d) Part 1: Split data

Here we will devide the data into a training set and a test set. Good rule of thumb is to use 80% of the data in the training set and 20 % in the test set. The the two data sets should be randomly sampled (shuffle=True), we also set he seed with random_state so that the split is reproducable. This is done with train_test_split() from sklearn, https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html. The syntax looks like 

train, test = train_test_split(dataframe, test_size=ratio_test_set, shuffle=True, random_state=42)

Why is it important that the data is shuffled when it is split, what could happen otherwise?

Text answer: Otherwise we could have a bias. Because they can have been grouped or made in order before. So we have to shuffle it to make sure we dont get that bias. 



In [205]:
# TODO: devide the data into train and test set. 
train,test = train_test_split(df, test_size = 0.2 , shuffle = True, random_state = 42)
#print(test["copium"])


## e) Part 1: Analyse data balance

The occurance can be retrevied with .value_counts() from a pandas date frame. Here get the occurance of copium in the samples. Is the dataset balanced?

In [206]:
# TODO: Print number of samples with copium and the number of samples without copium.
training_copium = train.value_counts("copium")
print(training_copium)
test_copium = test.value_counts("copium")
print(test_copium)
total_copium = df.value_counts("copium")
print(total_copium)


copium
0    687
1    123
Name: count, dtype: int64
copium
0    174
1     29
Name: count, dtype: int64
copium
0    861
1    152
Name: count, dtype: int64


Text answer: The number of samples with copium is a lot less than the samples without, therefore the dataset is not balanced.



## e) Part 2: Balance data

We will throughout this assignment test each method with the original dataset and a balanced dataset. We show how balanceing can be done by downsampling below. Your task is to create an upsampled balanced dataset. You only need to use the upsampled data set for the rest of the other part 2) exerices. 

What could be the reason for chosing either upsampled or downsampeled metohd to balance the dataset?


Text answer: 
We would like to upsample if the data we want is a small minority, if we upsamlpe with extra copies/variations of the minority you can make the model learn faster and at the same time balance the distribution. But with a risk of overfitting because of duplicates. We downsample if we feel that the minority has enough data to make it reliable while training. If we have a large dataset, downsampling can be good.


In [207]:
# TODO balance data det. 
# Seperate the data into something that contains copium and one that doesn't,
# can for example be done with df[df["copium"]==0] etc.
train_zero = train[train["copium"]==0]
train_one = train[train["copium"]==1]

# downsample majority
train_zero_downsampled = resample(train_zero,
                               n_samples=train_one.shape[0])

train_balanced_downsampled = pd.concat([train_one, train_zero_downsampled])
down_copium = train_balanced_downsampled.value_counts("copium")
#print(down_copium)



# TODO: upsample minority 
train_one_upsampled = resample(train_one, n_samples = train_zero.shape[0] )

train_balanced_upsampled = pd.concat([train_zero, train_one_upsampled])
up_copium = train_balanced_upsampled.value_counts("copium")


#print(train_zero) 687 rows
#print(train_balanced_upsampled) #810 rows
#print(train_balanced_upsampled["copium"][0:len(train_balanced_upsampled)/2])


#train_balanced_upsampled = 

# Exercise 3: Performance evaluation

Here we will define a class that later will be used for evaluation the performance of the classification methods. It will be used to log the accuracy, precsision and recalll. More information about precision and recall can be found at https://en.wikipedia.org/wiki/Precision_and_recall. 

Expliain why the different metics are usefull and why is not always accuarcy enough?

Text answer:
These all are there to understand a models performance. Precision measure the accuracy of positive predictions and thus it ratio, it focuses on the reliability of positive predictions. Recall is also known as sensitivity and measures the models ability to identity, it identify all positive instances.


In [208]:
class Classification_eval(object):
    def __init__(self):
        # counters 
        self.TP = 0 # correctly identified positive 
        self.FP = 0 # falsely identified positive 
        self.TN = 0 # correctly identified negative 
        self.FN = 0 # falsely identified negative 
    
    def update(self, pred, label):
        """
        pred - is the prediction will be either a 1 or 0. 
        label - is the correct answer, will be either a 1 or 0.
        """
        # TODO: add to one of the counters each time this function is called.
        if (pred == 1) and (label == 1):
            self.TP += 1
        elif  (pred == 0) and (label == 0):
            self.TN += 1
        elif (pred == 1) and (label == 0):
            self.FN += 1
        else:
            self.FP += 1
       

    def accuracy(self): 
        # returns the accuracy 
        if (self.TP + self.TN) == 0:
            return 0
        # TODO: calculate the accuracy.
        if (self.TP + self.TN + self.FN + self.FP) == 0:
            return 0
        
        accuracy = (self.TP + self.TN)/(self.TP + self.TN + self.FN + self.FP)

        return np.round(accuracy, 4)
    
    def precision(self): # percentage of the estimated positive that actually is positive
        if (self.TP + self.FP) == 0:
            return 0
        precision = self.TP/(self.TP + self.FN)    
        #precision = self.TP/(self.TP + self.FP)  #Faktiska formeln enligt föreläsning#
        
        # TODO: calculate the precision.
        return np.round(precision, 4)
    
    def recall(self): # percentage of correctly identified positive of the total positive
        if (self.TP + self.FN) == 0:
            return 0
        # TODO: calculate the recall.
        
        recall = self.TP/(self.TP + self.FP)
        #recall = self.TP/(self.TP + self.FN) #Faktiska formeln enligt föreläsning#
        return np.round(recall, 4)

In [209]:
# Test Classification_eval

test_eval = Classification_eval()
test_eval.update(1,1)
test_eval.update(1,1)
test_eval.update(1,1)
test_eval.update(0,1)
test_eval.update(1,0)
test_eval.update(1,0)
test_eval.update(0,0)
test_eval.update(0,0)
test_eval_ = np.allclose([test_eval.accuracy(), test_eval.precision(), test_eval.recall()], [0.625, 0.6, 0.75])
assert test_eval_
if test_eval_:
    print('Test successful')
else:
    print('Test failed')
    


Test successful


# Exercise 4: K- nearest neighbours

## a) Normalize

Here we will code our K-NN classifier, method 2.1 on page 21 in the book has the psudo code for K-NN. We will start with the data normalization, i.e. we will normalize the input data so that each feature has the same range in terms of max/min values. The min value can be found with data.min(), similarly for the max value. 

Why is it important that the data is normalized for the K-NN algorithm?

Text answer: We need to normalize in order to not neglect smaller distances while larger distances dominate.
You want to normalizide thus the KNN is distancedepending and if you dont normalizide the extreme values that can effect the distancecalculations. Through normalizing you make the algorithm more secure against these and make the datapoints to approximate same scale.


In [210]:
class Normalize(object):
    def __init__(self):
        self.min = None
        self.max = None
    
    def normalize(self, data):
        # normalize the data and return it. 
       
        return (data-self.min)/(self.max-self.min)
    
    def update_normalization(self, data):
        # Save the min and max values for each feature. This funciton is only used for the training data.
        self.min = data.min()
        self.max = data.max()
        return self.normalize(data)

    

In [211]:
class KNN(object):
    def __init__(self, k):
        self.features = None # normalized features from training data 
        self.labels = None # the corresponding labels (if there is copium)
        self.normalize = Normalize() # class instance for normalization
        self.k = k # the k value in k-nn algorithm.
        
        
        
    def fit(self, features, labels):

        normalized_features = self.normalize.update_normalization(features)
        features = normalized_features
        self.features = features
        self.labels = labels

    
    def predict(self, sample):
        #normailize
        sample = self.normalize.normalize(sample)
        #lists to hold k nearest points
        list_dist = np.ones(self.k)*np.inf
        list_prediction = np.zeros(self.k)
        #find k nearest points
        for i in range(self.features.shape[0]):
            dist = np.linalg.norm(self.features.iloc[i]-sample)
            if dist < np.max(list_dist):
                idx = np.argmax(list_dist)
                list_dist[idx] = dist
                list_prediction[idx] = self.labels.iloc[i]
        prediction = int(np.round(np.mean(list_prediction)))
                
        return prediction

In [212]:
# Test the KNN against known values. The test only covers small sample and there are hidden tests which will 
# tested after submitting. 

test_train = pd.read_csv('test.csv')
test_test = pd.read_csv('test2.csv')

train_labels = test_train['copium']
train_features = test_train.drop(columns='copium')
test_labels = test_test['copium']
test_features = test_test.drop(columns='copium')

knn = KNN(k=5)
knn.fit(train_features, train_labels)
pred_list = []

for i in range(test_features.shape[0]):
    pred = knn.predict(test_features.iloc[i])
    pred_list.append(pred)
val = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
if np.allclose(pred_list, val):
    print('Test successful')
else:
    print('Test failed')
    
#print((pred_list),"hej", (val)) 
    


Test successful


## c) part 1: Evaluate the K-NN 
Evaluate the K-NN and choose a suitable k value. 

In [246]:
train_labels = train['copium']
train_features = train.drop(columns='copium')

train_labels = train_balanced_upsampled['copium']
train_features = train_balanced_upsampled.drop(columns='copium')

y = test['copium']
x = test.drop(columns='copium')

# TODO, try differenent values of k. 
knn = KNN(k=50)
knn.fit(train_features, train_labels)

log = Classification_eval()
for i in range(x.shape[0]):
    pred = knn.predict(x.iloc[i])
    log.update(pred, y.iloc[i])

print('Accuarcy', log.accuracy())
print('Precision', log.precision())
print('Recall', log.recall())


Accuarcy 0.867
Precision 0.5263
Recall 0.6897


Try some differnet values of k and just looking at these resutlts would the klassifier work well for all k? 

Answer:

| k | Accuracy | Precision | Recall | 
| --- | --- | --- | --- |
| 1 |0.867 | 0.5556| 0.3448|  
| 5 |0.7783 | 0.2759| 0.25|   
| 20 | 0.8621| 0| 0|   
| 50 |0.8571 |0 | 0|  


## c) part 2, balanced data. 
Now try with the upsampled balanced data, test for the same k values as in part 1. Are the results different and whitch one would identify more copium?  

Answer:


| k | Accuracy | Precision | Recall | 
| --- | --- | --- | --- |
| 1 | 0.867| 0.5556 |0.3448 |  
| 5 |0.8227 |0.4186 | 0.6207|   
| 20 |0.9064 | 0.6316| 0.8276|   
| 50 | 0.9015| 0.6286| 0.7586|

## d) kNN with Sklearn
Typically one would not implement these algorithms by hand for each time you need them. Here we show how the kNN algorithm from sklearn can be used. Note, it does not normalize the features by default. The advantanges with the sklearn version is in computational cost as they utilize clever tree structures to minimize the search problem. 

A quick note: Even if ready packages exist, it is still very valuable to understand the principels behind the algorithms. For example when the learning doesn't work as well as you liked (happens more often then not...), you can reason to why and modify the problem or method into something that does work.  

In [214]:
# Set up data used for fitting the model
train_labels = train['copium']
train_features = train.drop(columns='copium')

# with the off the shelf model we need normalize the data before 
normalize = Normalize()
data_features = normalize.update_normalization(train_features)

# The test data
y = test['copium']
x = test.drop(columns='copium')

# Normalize the test data
x_norm = normalize.normalize(x)

# Set up an knn with sklearn 
knn_sklearn = KNeighborsClassifier(n_neighbors=5)
knn_sklearn.fit(data_features.values, train_labels.values)

# Set up our own for comparison 
knn = KNN(k=5)
knn.fit(train_features, train_labels)

# Test both for the 20 first values
for i in range(20):
    pred = knn.predict(x.iloc[i])
    pred_sklearn = knn_sklearn.predict([x_norm.iloc[i]])
    print('Prediciton from our model', pred, ' and from sklearn ',pred_sklearn)


Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 1  and from sklearn  [1]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  and from sklearn  [0]
Prediciton from our model 0  an

# Exercise 5: Tree based classifier

In this exercise we will implement a thee based classifier. The implementation will be a class representing a node and will recursivly create child node until the entire tree structure is built. The general step for learning a tree structure from learning data is as follows:

1. Initiate the the root node.
2. Call the learn function with the learning data. 

Inside the learn function

3. Check if all the data has the same classification, and check if the data set is smaller then min_node_size. If any of these are true, set the node as a leaf node and return. 
4. Find the spliting point in the data that returns the highest gini value, be careful to not use the label which we want to classify. 
5. Split the dataset, one is usually called head and the other tail. Save the splitting parameter and which value was used for the splitting. 
6. Check the prediction of the splited data sets (majority copium or not) and create two child nodes, one for each of the splited dataset. 
7. Call the the learn function of the child node with the corresponding split data and appdend the child node. After that return. 

The above will build the tree structure. To use the tree for prediction simply call the predict function of the root node with a data sample. In the predict function the following steps should happen.

1. Check if the current node is a lead node, if it is simply return the self.prediction of that node. 
2. If it is not a leaf node then check the value of the splitting parameter in the data sample and compare with the splitting value. 
3. Call the predict function of the correspoinding child node with the data sample and return the answer.

This will recurivly traverse the tree until a leaf node containing the prediction is reached. 

 


## 5 a) Find split point

First we define a function that can find the splitting point for a parameter with the highest gini value. 

The gini value can be described as:

If $\Gamma$ contains the set of all labeles, then $\Gamma(x < c)$ would be all labels that belong to the criteria $x < c$, where $x$ is a feature or parameter and $c$ is the splitting criteia (split_value).  


For example: 

If $x$ is iron_rate and our splitting critera is $c = 0.4$, then $\Gamma(x < c)$ would be a set of all the label from samples with $\textrm{iron_rate} < 0.4$.


Then we can define:

$v_1 = mean(\Gamma(x < c))$

$v_2 = mean(\Gamma(x \geq c))$

$s_1 = v_1^2 + (1-v_1)^2$

$s_2 = v_2^2 + (1-v_2)^2$

We define $len(\Gamma)$ to give the number of samples in the dataset, then the weighted gini value is:

$s = \frac{len(\Gamma(x_i < c))}{len(\Gamma)}*s_1 + \frac{len(\Gamma(x_i \geq c))}{len(\Gamma)}*s_2$

The goal is to find a criteria $c$ that maximizes the gini value $s$. The easies is to order the data set after parameter, e.g. we could order all sample depending on how much iron_rate they have, then we can try to split the data between every data point in a for loop and save the one that has the highest gini value. The critieria c would be the average between two data points that are sorted, i.e. $c_i = \frac{x_{i-1} + x_i}{2}$. 

In [215]:
def find_split_point(data, label, parameter):
    """
    data - all the data we want to split, our (gamma)
    label - the parameter we want to classify. 
    parameter - the parameter we want to check for, our x_i
    -----------
    retrun:
    split_value - the spliting value, our c. 
    gini_value - the gini value for the best c.
    df_head - the data frame belonging to x_i < c
    df_tail - the data frame belonging to x_i => c
    """
    # Beging by sorting the data after the paramter. 
    sorted_data = data.sort_values(by=parameter)    #sorted by ground_density, ground_density is our x_i
    sorted_label = sorted_data[label]
    
    # TODO loop through all the split points in the sorted data and find 
    # the best gini_value (s) and split_value (c).  
    n = sorted_data.shape[0]
    new_gini_value = 0
    split_value = 0

    for i in range(1,n):
  
        #kollar bara på y värden då x är på båda sidor av splitkravet
        label_head = sorted_label[:i]      #can slice, but can't access single value for some reason
        label_tail = sorted_label[i:]
      
        v1 = np.sum(label_head)/label_head.shape[0]
        v2 = np.sum(label_tail)/label_tail.shape[0]
   
        s1 = v1 ** 2 +(1 - v1) ** 2
        s2 = v2 ** 2 + (1 - v2) ** 2
        
        gini_value = label_head.shape[0]/n * s1 + label_tail.shape[0]/n * s2
        if gini_value > new_gini_value:
            new_gini_value = gini_value
            split_value = (sorted_data[parameter].iloc[i-1] + sorted_data[parameter].iloc[i]) /2

    df_head = sorted_data[sorted_data[parameter] < split_value]
    df_tail = sorted_data[sorted_data[parameter] >= split_value]
   
    return split_value, new_gini_value, df_head, df_tail 
    
    

In [216]:
# Test for find_split_point function, the test does not test everything and more hidden tests are used 
# during grading. 
df_test = pd.read_csv('test.csv')
split_value, gini_value, df_head, df_tail = find_split_point(df_test, 'copium', 'ground_density')
if split_value == 1.8762144694722296 and gini_value == 0.7234487734487733:
    print('Test successful')
else:
    print('Test failed')
    
    

Test successful


## 5 b) Tree Node

Here we will code a tree node class that can recursivly call it self. 


In [217]:
class TreeNode():
    def __init__(self, prediction=None):
        self.split_value = None # the splitting value (c)
        self.split_parameter = None # what feature where uesd for the split (x_i)
        self.child_nodes = [] # list that contains two child nodes, if not leaf_node
        self.leaf_node = 0 # is this leaf_node (0=no, 1=yes)
        self.prediction = prediction # classification made in this node.
        
    def predict(self, data):
        # TODO: To make a prediction we need to traverse the tree recursivly down to a leaf node and return 
        # the prediction. 
        # step 1: check if this is a leaf node, if it is then return prediction of this node, 
        if self.leaf_node == 1:
            return self.prediction
        # otherwise contine with step 2.
        # step 2: Compare the input data with the splitting criteria.
        if data[self.split_parameter] < self.split_value:
            #self.child_nodes[left_childnode, right_child_node]
            return self.child_nodes[0].predict(data)       #traversing recursively
        else:
            return self.child_nodes[1].predict(data)

        # step 3: Call the prediction function in the corresponding child_node and return the prediction.
         
        
            
    def learn(self, data, label, min_node_size):
        """
        data - the training data
        label - the parameter we want to classify, i.e. "copium"
        min_node_size - number of data points in a node for it to become a leaf node. 
        """
        # TODO: wirte the learning function.
        new_split_value = 0 
        new_gini_value = 0
        new_df_head = 0
        new_df_tail = 0
      
        # Step 1: check if this node should be turned into a leaf node, this happens if the data set is smaller 
        # then the min_node_size or if the data labels are homogenious. 
        if data[label].mean() == 1 or data[label].mean() == 0 or (len(data) < min_node_size):
            self.leaf_node = 1
            return
            #return self.prediction
        
      
  
        
    
        
        # Step 2: Loop over all features and get the best gini and split_value for each feature. 
        # Save the best split value and split_paramter and the two new data frames 
        # corresponding to the split [df_head, df_tail] (these will be data frames for the child nodes).  
        #for feature in data.columns:
        for feature in data.drop(columns=label).columns:
           # print(feature)      #10 features
            split_value1, gini_value, df_head, df_tail = find_split_point(data, label, feature) 
            if gini_value > new_gini_value:
                new_gini_value = gini_value
                new_split_value = split_value1
                new_df_head = df_head   #dataframe for childnodes
                new_df_tail = df_tail   #dataframe for childnodes
                self.split_parameter = feature
      
       # self.child_nodes = [df_head, df_tail]
        self.split_value = new_split_value
       
        majority_head = int(np.round(np.mean(new_df_head[label]))) #0
        majority_tail = int(np.round(np.mean(new_df_tail[label]))) #1
  
        #print(int(df_tail))
        #print(data.columns)
      
        child1 = TreeNode(prediction = majority_head)
        child1.learn(new_df_head, label, min_node_size)
        child2 = TreeNode(prediction = majority_tail)
        child2.learn(new_df_tail, label, min_node_size)

        self.child_nodes.append(child1)
        self.child_nodes.append(child2)
        #print(self.child_nodes)
        
        
        # Step 3: Calculate the majority vote prediction of the two data frames. 
        # Create two child nodes, (e.g. child = TreeNode(prediction=1) for prediction copium) and and call the 
        # learn function with the corresponding dataframe. 
        
        # Step 4: append the the child node to the self.child_nodes. It should be in the order 
        # of the child node correspoinding to [df_head, df_tail].
        
        
        

        return

In [218]:
# Test to give indicaion if the TreeNode class is correct.

df_test = pd.read_csv('test.csv')
tree = TreeNode() # create root node
# learn the tree structure
tree.learn(df_test, "copium", min_node_size=4)

pred_list = []
pred_answer = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
               0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 
               0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
               1, 0, 1, 0, 1, 0, 0, 0, 0]

for i in range(df_test.shape[0]):
    pred = tree.predict(df_test.iloc[i])
    pred_list.append(pred)
    
if np.allclose(pred_list, pred_answer):
    print('Test successful')
else:
    print('Test failed')
print(pred_list,"\n", "!" ,pred_answer)

Test successful
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0] 
 ! [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0]


##  Train the Tree Classifier

In [244]:
tree = TreeNode()
tree.learn(train, "copium", min_node_size=1)


## Test Classifier

In [245]:
y = test['copium']
x = test.drop(columns='copium')
log = Classification_eval()

for i in range(x.shape[0]):
    pred = tree.predict(x.iloc[i])
    log.update(pred, y.iloc[i])
        
print('accuarcy', log.accuracy())
print('precision', log.precision())
print('recall', log.recall())

accuarcy 0.8966
precision 0.6429
recall 0.6207


## 5 c) part 1

Use the non balanced data to learn the tree classifier and test the classifier for some different min_node_size values.

Answer:



| min_node_size | Accuracy | Precision | Recall | 
| --- | --- | --- | --- |
| 1  | 0.8966 | 0.6429 | 0.6207 |  
| 10 | 0.8916 | 0.6207 | 0.6207 |   
| 20 | 0.8867 | 0.6364 | 0.4828 |   
| 50 | 0.8818 | 0.6471 | 0.3793 |  

## 5 c) part 2, Try with balanced data

Use the balanced data to learn the tree classifier and test the classifier for some different min_node_size values.

Answer:


| min_node_size | Accuracy | Precision | Recall | 
| --- | --- | --- | --- |
| 1  |  0.8424| 0.4595 | 0.5862 |  
| 10 | 0.8522 | 0.4857 | 0.5862 |   
| 20 | 0.8571 | 0.5 | 0.5862 |   
| 50 |0.8522  | 0.4865 | 0.6207 |

## d) Tree classifier with sklearn 

Here we compare our tree classifier to the one implemented in sklearn. It is important to note that there can be differences between them, unlike our implementation the sklearn has some randomness in the traing and does not produce the same results each time. But overall they give simlar results. 

In [232]:
# Set up data used for fitting the model
train_labels = train['copium']
train_features = train.drop(columns='copium')

# Decition tree or classifier from sklearn 
tree_sklearn = DecisionTreeClassifier(min_samples_split=1.0, splitter='best', 
                             criterion='gini', min_samples_leaf=1)
# Train the sklearn tree classifer
tree_sklearn.fit(train_features.values, train_labels.values)

# Train our tree classifier 
tree = TreeNode() # create root node
tree.learn(train, "copium", min_node_size=1)

# Test both and compare
y = test['copium']
x = test.drop(columns='copium')
# classify the data
for i in range(30):
    pred_sklearn = tree_sklearn.predict([x.iloc[i]])
    pred = tree.predict(x.iloc[i])

    print('sklearn pred ',pred_sklearn, 'our tree pred ', pred)

sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [1] our tree pred  1
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [1] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [1] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [1] our tree pred  1
sklearn pred  [0] our tree pred  0
sklearn pred  [0] our tree pred  0
sklearn pred  [0] ou

## Exersice 6: Balance data analysis

1. It there a difference between that results of the balanced and non-balanced data for the tree classifier?
2. Is there a difference between the K-NN vs tree classifier?
3. Which one would you expect to be a faster classifier?

Answer:
1.  First of all, we should've swapped placed with recall and precision, we have them like this to pass the test.
Yes. The upsampled is more consistent. We can see that the accuracy/precision and recall are less dependant on the min node size of our leaf nodes. In the non-balanced data we get less accuracy the larger we set our min node size to. 

2. The difference is that KNN gets better the more neighbours we are looking while upsampling, while if we decrease the min node size of our leaves in tree we get higher recall and precision.

3. The tree classifier should be faster, especially scaling with larger data sizes.



## Exersice 7: Deployment (Optional)

Here we will try the learned classifiers on a larger map. Make sure that the last run version of K-NN and tree have good parameters i.e. k and min_node_size values. Which one found more copium?


In [ ]:
env = Environment(map_type=2, fps=5, resolution=(1000, 1000))

try:
    sensor_properties = env.get_sensor_properties()
    sensor_sample = dict()
    for key in sensor_properties:
        sensor_sample[key] = [0]

    log_knn = Classification_eval()
    log_tree = Classification_eval()


    for i in range(500):
        action = breadth_first_search(actor=env.get_actor(), max_depth=3, action_space=env.get_action_space())
        env.step(action)
        if env.get_sensor_readings() is not None:
            sensor_readings = env.get_sensor_readings()
            for key in sensor_readings:
                sensor_sample[key][0] = sensor_readings[key]
            sensor_sample_df = pd.DataFrame(sensor_sample)
            log_knn.update(knn.predict(sensor_sample_df.iloc[0]), env.get_ground_truth())
            log_tree.update(tree.predict(sensor_sample_df.iloc[0]), env.get_ground_truth())
            env.plt_acc.update_acc(log_tree.accuracy(), log_knn.accuracy())
        env.render()

    env.exit()

    print("K-NN accuracy ", log_knn.accuracy(), "Tree accuracy", log_tree.accuracy())
    print("Number of copium deposits found, K-NN:", log_knn.TP, " Tree:", log_tree.TP)
except:
    env.exit()



